### ISSUE TITLE
Inconsistent missing argument error

### ISSUE THREAD

**Issue by jverre** [openscoring/openscoring#22]
I have created a PMML that runs on one openscoring instance but not on another. On one openscoring instance, the model returns the expected result while on the other it returns:
```
org.jpmml.evaluator.FunctionException (at or around line 21): Missing arguments
at org.jpmml.evaluator.functions.AbstractFunction.checkArguments(AbstractFunction.java:60) 
at org.jpmml.evaluator.functions.AbstractFunction.checkArguments(AbstractFunction.java:40) 
at org.jpmml.evaluator.functions.EqualityFunction.evaluate(EqualityFunction.java:40) 
at org.jpmml.evaluator.FunctionUtil.evaluate(FunctionUtil.java:47) 
at org.jpmml.evaluator.ExpressionUtil.evaluateApply(ExpressionUtil.java:439) 
at org.jpmml.evaluator.ExpressionUtil.evaluateExpression(ExpressionUtil.java:106) 
at org.jpmml.evaluator.ExpressionUtil.evaluate(ExpressionUtil.java:66) 
at org.jpmml.evaluator.ExpressionUtil.evaluateApply(ExpressionUtil.java:356) 
at org.jpmml.evaluator.ExpressionUtil.evaluateExpression(ExpressionUtil.java:106) 
at org.jpmml.evaluator.ExpressionUtil.evaluate(ExpressionUtil.java:66)
at org.jpmml.evaluator.OutputUtil.evaluate(OutputUtil.java:203) 
at org.jpmml.evaluator.rule_set.RuleSetModelEvaluator.evaluate(RuleSetModelEvaluator.java:112) 
at org.jpmml.evaluator.ModelEvaluator.evaluate(ModelEvaluator.java:373) 
at org.openscoring.service.ModelResource.evaluate(ModelResource.java:570) 
at org.openscoring.service.ModelResource.doEvaluate(ModelResource.java:418) 
at org.openscoring.service.ModelResource.evaluate(ModelResource.java:266) 
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) 
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) 
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) 
at java.lang.reflect.Method.invoke(Method.java:498) 
at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81) 
at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:144) 
at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:161) 
at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$TypeOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:205) 
at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:99) 
at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:389) 
at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:347) 
at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:102) 
at org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:326) 
at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271) 
at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267) 
at org.glassfish.jersey.internal.Errors.process(Errors.java:315) 
at org.glassfish.jersey.internal.Errors.process(Errors.java:297) 
at org.glassfish.jersey.internal.Errors.process(Errors.java:267) 
at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317) 
at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:305) 
at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1154) 
at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:473) 
at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:427) 
at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:388) 
at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:341) 
at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:228) 
at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:812) 
at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:587) 
at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127) 
at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515) 
at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061) 
at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) 
at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:215) 
at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97) 
at org.eclipse.jetty.server.Server.handle(Server.java:499) 
at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311) 
at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:258) 
at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544) 
at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635) 
at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555) 
at java.lang.Thread.run(Thread.java:745)
```

This issue seems related to the output field and whether or not the predicted field is included.

I have attached a test PMML that exhibits the behaviour described above.

Any ideas why openscoring would have two different behaviours for this one PMML ?

PMML:
[test.txt](https://github.com/openscoring/openscoring/files/1008237/test.txt)

**Comment by vruusmann** [comment]
As the exception shows, the failing construct is the Apply element on lines 21 -- 24:
```xml
<Apply function="equal">
	<FieldRef field="raw_result"/>
	<Constant dataType="string">noRuleFound</Constant> 
</Apply>
```

According to the PMML specification, [the "equal" built-in function](http://dmg.org/pmml/v4-2-1/BuiltinFunctions.html#boolean2) takes two arguments, which have to be both non-null. In your case, one (or both) of the arguments is null.

Most likely, the missing argument is `<FieldRef field="raw_result">`, which references the prediction of the rule set model. This model is very naive, and it's easy to see that it will return a null prediction every time when the value of the `value` input field is something other than `0`.

In conclusion, the scoring will succeed if you pass `value=0` as input, and fail otherwise. The Openscoring REST service is working properly.

**Comment by jverre** [comment]
Given that I have provided a default score, shouldn't the raw_result field (which is the predicted field) never be null ?

**Comment by vruusmann** [comment]
The contents of the "result" output field, expressed as Java code:
```Java
if(raw_result == "noRuleFound"){
  return default_value;
} else {
  return raw_result;
}
```

It can be seen that the "if" condition has to execute successfully (and produce a `true` value) before the `default_value` could be selected. You have a situation where the "if" condition fails (by throwing a `NullPointerException`).

**Comment by vruusmann** [comment]
As a workaround, you can check the "missingness" state of the `raw_result` field:
```xml
<Apply function="if">
	<Apply function="or">
		<Apply function="isMissing">
			<FieldRef field="raw_result"/>
		</Apply>
		<Apply function="equal">
			<FieldRef field="raw_result"/>
			<Constant>noRuleFound</Constant>
		</Apply>
	</Apply>
</Apply>
```

**Comment by jverre** [comment]
Thanks for the help !
I gave the two workarounds a try but they didn't actually solve my issue unfortunately.

The issue I have is that the code:
* Will run on machine 1 but not machine 2 if I use:
``` <Output>
         <OutputField name="ruleId" optype="continuous" dataType="double" feature="entityId"/> 
          <OutputField name="result" feature = "transformedValue">
            <Apply function="if">
              <Apply function="equal">
                <FieldRef field="raw_result"/>
                <Constant dataType="string">noRuleFound</Constant> 
              </Apply>
              <FieldRef field="default_value"/>
              <FieldRef field="raw_result"/>  
            </Apply>
          </OutputField>
        </Output>```
* Will run on machine 2 but not machine 1 if I use:
``` <Output>
         <OutputField name="raw_result" optype="categorical" dataType="string" feature="predictedValue"/> 
         <OutputField name="ruleId" optype="continuous" dataType="double" feature="entityId"/> 
         <OutputField name="result" feature = "transformedValue">
            <Apply function="if">
              <Apply function="equal">
                <FieldRef field="raw_result"/>
                <Constant dataType="string">noRuleFound</Constant> 
              </Apply>
              <FieldRef field="default_value"/>
              <FieldRef field="raw_result"/>  
            </Apply>
          </OutputField>
        </Output>```

Both machines use the same version of openscoring.

**Comment by vruusmann** [comment]
> Both machines use the same version of openscoring.

Are you absolutely positively sure about that?

The difference between two code examples is this:
```
<OutputField name="raw_result" optype="categorical" dataType="string" feature="predictedValue"/> 
```

Newer versions of openscoring (1.3.X) will fail if the prediction hasn't been imported into output scope using a specialized output field (`<OutputField feature="predictedValue"/>`), whereas older version (1.2.X) silently work around it.

Therefore, it must be the case that you use older version on machine 1, and newer version on machine 2. Whateever the case, please upgrade to the latest version of openscoring on both machines to remove all doubt.

**Comment by jverre** [comment]
Thanks, this must be the reason. I will investigate tomorrow morning.

Thanks again for all the support !
